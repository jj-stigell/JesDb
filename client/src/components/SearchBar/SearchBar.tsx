import React from 'react'
import { HiOutlineSearch, HiX } from 'react-icons/hi'

interface ISearchBar {
  keyword: string
  disabled: boolean
  loading: boolean
  inputRef?: React.RefObject<HTMLInputElement>
  setKeyword: (q: string) => void
}

export default function SearchBar ({
  keyword,
  disabled,
  loading,
  inputRef,
  setKeyword
}: ISearchBar): React.JSX.Element {
  const width = window.innerWidth
  const placeholder = width > 649
    ? 'Search with Japanese or English, e.g, 頑張る...'
    : 'Search e.g., 頑張る...'

  const handleChange = (event: any): void => {
    setKeyword(event.target.value)
  }

  const handleFocus = (): void => {
    if (inputRef?.current !== null) {
      inputRef?.current.select()
    }
  }

  const reset = (): void => {
    setKeyword('')
    if (inputRef?.current !== null) {
      inputRef?.current.focus()
    }
  }

  return (
    <div className="w-full">
      <label
        htmlFor="search"
        id="search-label"
        className="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white"
      >
        Search
      </label>
      <div className="relative" role="search">
        <input
          id="search"
          type="search"
          ref={inputRef}
          required
          autoFocus
          value={keyword}
          disabled={disabled}
          aria-labelledby="search-label"
          aria-busy={loading}
          onFocus={handleFocus}
          onChange={handleChange}
          placeholder={placeholder}
          className={`
            block w-full p-3 ps-6 md:ps-10 text-lg text-gray-900 border border-2 border-gray-300
            rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700
            dark:border-gray-600 dark:placeholder-gray-300 dark:text-white
            dark:focus:ring-jesdark-500 dark:focus:border-jesdark-500
          `}
        />
        <button
          type="submit"
          disabled={disabled}
          aria-label="Search"
          data-testid="search-button"
          className={`
            text-white absolute end-2.5 bottom-2.5  focus:ring-4 focus:outline-none
            focus:ring-blue-300 font-medium rounded-lg text-sm px-4 py-2
            ${disabled
              ? 'bg-blue-400 dark:bg-jesdark-700'
              : 'dark:bg-jesdark-500 bg-blue-600 hover:bg-blue-800 dark:hover:bg-jesdark-700'
            }
          `}
        >
          {loading
            ? (
                <div
                  role="status"
                  aria-live="polite"
                  data-testid="load-spinner"
                  className={`
                  h-5 w-5 animate-spin rounded-full border-4 border-solid border-current
                  border-e-transparent motion-reduce:animate-[spin_1.5s_linear_infinite]
                  `
                  }
                />
              )
            : (
                <HiOutlineSearch className="h-5 w-5" />
              )}
        </button>
        <button
          type="button"
          onClick={reset}
          disabled={disabled}
          aria-label="Reset input"
          data-testid="reset-button"
          className={`
            text-grey absolute end-16 bottom-2.5 font-medium
            rounded-lg text-sm px-4 py-2 dark:text-white
          `}
        >
          <HiX className="h-5 w-5" />
        </button>
      </div>
    </div>
  )
}
