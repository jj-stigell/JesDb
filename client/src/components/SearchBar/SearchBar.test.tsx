import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import SearchBar from './SearchBar'

describe('SearchBar Component', () => {
  const setKeyword = jest.fn()

  beforeEach(() => {
    setKeyword.mockClear()
  })

  it('renders with correct placeholder for large screens', () => {
    // Mock window.innerWidth
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 650 })
    const { asFragment } =
      render(<SearchBar keyword="" loading={false} disabled={false} setKeyword={setKeyword} />)

    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByPlaceholderText('Search with Japanese or English, e.g, 頑張る...'))
      .toBeInTheDocument()
  })

  it('renders with correct placeholder for small screens', () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 649 })
    const { asFragment } =
      render(<SearchBar keyword="" loading={false} disabled={false} setKeyword={setKeyword} />)

    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByPlaceholderText('Search e.g., 頑張る...')).toBeInTheDocument()
  })

  it('updates keyword on input change', () => {
    render(<SearchBar keyword="" loading={false} disabled={false} setKeyword={setKeyword} />)
    fireEvent.change(screen.getByRole('searchbox'), { target: { value: 'test' } })

    expect(setKeyword).toHaveBeenCalledWith('test')
  })

  it('resets keyword when reset button is clicked', () => {
    render(<SearchBar keyword="test" loading={false} disabled={false} setKeyword={setKeyword} />)
    fireEvent.click(screen.getByTestId('reset-button'))

    expect(setKeyword).toHaveBeenCalledWith('')
  })

  it('disables input and buttons when disabled', () => {
    render(<SearchBar keyword="test" loading={false} disabled={true} setKeyword={setKeyword} />)
    const input = screen.getByRole('searchbox')
    const searchButton = screen.getByTestId('search-button')
    const resetButton = screen.getByTestId('reset-button')

    expect(input).toBeDisabled()
    expect(searchButton).toBeDisabled()
    expect(resetButton).toBeDisabled()
  })

  it('shows spinner when loading', () => {
    render(<SearchBar keyword="test" loading={true} disabled={false} setKeyword={setKeyword} />)
    expect(screen.getByTestId('load-spinner')).toBeInTheDocument()
  })

  it('matches snapshot with keyword set', () => {
    const { asFragment } =
      render(<SearchBar keyword="test" loading={false} disabled={false} setKeyword={setKeyword} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
