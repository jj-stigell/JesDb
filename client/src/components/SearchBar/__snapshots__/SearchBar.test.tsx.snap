// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component matches snapshot with keyword set 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white"
      for="search"
      id="search-label"
    >
      Search
    </label>
    <div
      class="relative"
      role="search"
    >
      <input
        aria-busy="false"
        aria-labelledby="search-label"
        class="
            block w-full p-3 ps-6 md:ps-10 text-lg text-gray-900 border border-2 border-gray-300
            rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700
            dark:border-gray-600 dark:placeholder-gray-300 dark:text-white
            dark:focus:ring-jesdark-500 dark:focus:border-jesdark-500
          "
        id="search"
        placeholder="Search e.g., 頑張る..."
        required=""
        type="search"
        value="test"
      />
      <button
        aria-label="Search"
        class="
            text-white absolute end-2.5 bottom-2.5  focus:ring-4 focus:outline-none
            focus:ring-blue-300 font-medium rounded-lg text-sm px-4 py-2
            dark:bg-jesdark-500 bg-blue-600 hover:bg-blue-800 dark:hover:bg-jesdark-700
          "
        data-testid="search-button"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="h-5 w-5"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <button
        aria-label="Reset input"
        class="
            text-grey absolute end-16 bottom-2.5 font-medium
            rounded-lg text-sm px-4 py-2 dark:text-white
          "
        data-testid="reset-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="h-5 w-5"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SearchBar Component renders with correct placeholder for large screens 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white"
      for="search"
      id="search-label"
    >
      Search
    </label>
    <div
      class="relative"
      role="search"
    >
      <input
        aria-busy="false"
        aria-labelledby="search-label"
        class="
            block w-full p-3 ps-6 md:ps-10 text-lg text-gray-900 border border-2 border-gray-300
            rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700
            dark:border-gray-600 dark:placeholder-gray-300 dark:text-white
            dark:focus:ring-jesdark-500 dark:focus:border-jesdark-500
          "
        id="search"
        placeholder="Search with Japanese or English, e.g, 頑張る..."
        required=""
        type="search"
        value=""
      />
      <button
        aria-label="Search"
        class="
            text-white absolute end-2.5 bottom-2.5  focus:ring-4 focus:outline-none
            focus:ring-blue-300 font-medium rounded-lg text-sm px-4 py-2
            dark:bg-jesdark-500 bg-blue-600 hover:bg-blue-800 dark:hover:bg-jesdark-700
          "
        data-testid="search-button"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="h-5 w-5"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <button
        aria-label="Reset input"
        class="
            text-grey absolute end-16 bottom-2.5 font-medium
            rounded-lg text-sm px-4 py-2 dark:text-white
          "
        data-testid="reset-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="h-5 w-5"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SearchBar Component renders with correct placeholder for small screens 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white"
      for="search"
      id="search-label"
    >
      Search
    </label>
    <div
      class="relative"
      role="search"
    >
      <input
        aria-busy="false"
        aria-labelledby="search-label"
        class="
            block w-full p-3 ps-6 md:ps-10 text-lg text-gray-900 border border-2 border-gray-300
            rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700
            dark:border-gray-600 dark:placeholder-gray-300 dark:text-white
            dark:focus:ring-jesdark-500 dark:focus:border-jesdark-500
          "
        id="search"
        placeholder="Search e.g., 頑張る..."
        required=""
        type="search"
        value=""
      />
      <button
        aria-label="Search"
        class="
            text-white absolute end-2.5 bottom-2.5  focus:ring-4 focus:outline-none
            focus:ring-blue-300 font-medium rounded-lg text-sm px-4 py-2
            dark:bg-jesdark-500 bg-blue-600 hover:bg-blue-800 dark:hover:bg-jesdark-700
          "
        data-testid="search-button"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="h-5 w-5"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <button
        aria-label="Reset input"
        class="
            text-grey absolute end-16 bottom-2.5 font-medium
            rounded-lg text-sm px-4 py-2 dark:text-white
          "
        data-testid="reset-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="h-5 w-5"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
